demos/coreMQTT/mqtt_demo_serializer.c

```diff
@@ -0,0 +1,1620 @@
+}
+/*-----------------------------------------------------------*/
+
+static int32_t prvTransportRecv( const NetworkContext_t * pxContext,
+                                 void * pvBuffer,
+                                 size_t xBytesToRecv )
+{
+    int32_t lResult;
+
+    configASSERT( pxContext != NULL );
+
+    /* Receive upto xBytesToRecv from network. */
+    lResult = ( int32_t ) SOCKETS_Recv( ( Socket_t ) pxContext->xTcpSocket,
+                                        pvBuffer,
+                                        xBytesToRecv,
+                                        0 );
+
+    return lResult;
+}
+/*-----------------------------------------------------------*/
+
+static uint16_t prvGetNextPacketIdentifier()
+{
+    static uint16_t usPacketId = 0;
+
+    usPacketId++;
+
+    /* Since 0 is invalid packet identifier value,
+     * take care of it when it rolls over */
+    if( usPacketId == 0 )
+    {
+        usPacketId = 1;
+    }
+
+    return usPacketId;
+}
+/*-----------------------------------------------------------*/
+
+static Socket_t prvCreateTCPConnectionToBroker( void )
+{
+    Socket_t xMQTTSocket = ( Socket_t ) SOCKETS_INVALID_SOCKET;
+    uint32_t ulBrokerIPAddress;
+    SocketsSockaddr_t xBrokerAddress;
+    int32_t lSocketStatus = SOCKETS_ERROR_NONE;
+
+    /* This is the socket used to connect to the MQTT broker. */
+    xMQTTSocket = SOCKETS_Socket( SOCKETS_AF_INET,
+                                  SOCKETS_SOCK_STREAM,
+                                  SOCKETS_IPPROTO_TCP );
+
+    if( xMQTTSocket != SOCKETS_INVALID_SOCKET )
+    {
+        /* Socket was created. Locate the endpoint, then connect to the MQTT broker. */
+        ulBrokerIPAddress = SOCKETS_GetHostByName( democonfigMQTT_BROKER_ENDPOINT );
+
+        if( ulBrokerIPAddress != 0 )
+        {
+            xBrokerAddress.usPort = SOCKETS_htons( democonfigMQTT_BROKER_PORT );
+            xBrokerAddress.ulAddress = ulBrokerIPAddress;
+
+            lSocketStatus = SOCKETS_Connect( xMQTTSocket, &xBrokerAddress, sizeof( xBrokerAddress ) );
+
+            if( lSocketStatus != SOCKETS_ERROR_NONE )
+            {
+                LogError( ( "Could not connect to MQTT broker %s: SOCKETS_Connect() failed. SocketStatus=%d.",
+                            democonfigMQTT_BROKER_ENDPOINT,
+                            lSocketStatus ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Could not connect to broker: DNS resolution failed: Broker=%s.",
+                        democonfigMQTT_BROKER_ENDPOINT ) );
+        }
+    }
+    else
+    {
+        LogError( ( "Could not create TCP socket." ) );
+    }
+
+    /* If the socket was created but the connection was not successful then delete
+     * the socket again. */
+    if( lSocketStatus != SOCKETS_ERROR_NONE )
+    {
+        if( xMQTTSocket != SOCKETS_INVALID_SOCKET )
+        {
+            SOCKETS_Close( xMQTTSocket );
+            xMQTTSocket = SOCKETS_INVALID_SOCKET;
+        }
```

Comment:
```suggestion
            if( SOCKETS_Connect( xMQTTSocket, &xBrokerAddress, sizeof( xBrokerAddress ) ) != SOCKETS_ERROR_NONE )
            {
                LogError( ( "Could not connect to MQTT broker %s: SOCKETS_Connect() failed. SocketStatus=%d.",
                            democonfigMQTT_BROKER_ENDPOINT,
                            lSocketStatus ) );
                            
                /* Clean up the created socket. */            
                SOCKETS_Close( xMQTTSocket );
                xMQTTSocket = SOCKETS_INVALID_SOCKET;           
            }
        }
        else
        {
            LogError( ( "Could not connect to broker: DNS resolution failed: Broker=%s.",
                        democonfigMQTT_BROKER_ENDPOINT ) );
        }
    }
    else
    {
        LogError( ( "Could not create TCP socket." ) );
    }
```

Reply:
Then it would have to go into tow error cases.